File: php/src/detectors/ShellshockDetector.php
<?php

namespace FireWTWall\Detectors;

use FireWTWall\Request;

/**
 * Detects Shellshock (CVE-2014-6271 / CVE-2014-7169) exploitation attempts.
 *
 * Shellshock payloads are delivered via HTTP headers, so this detector scans
 * ALL HTTP headers by iterating $_SERVER keys that begin with HTTP_.
 *
 * Requires PHP >= 8.0
 */
class ShellshockDetector
{
	private static array $rules = [
		['id' => 'shellshock-func',    'pattern' => '/\(\s*\)\s*\{\s*[^}]*\}\s*;/',    'severity' => 'critical'],
		['id' => 'shellshock-env-cmd', 'pattern' => '/\(\s*\)\s*\{\s*:;\s*\}\s*;/',    'severity' => 'critical'],
	];

	/**
	 * Scan all HTTP request headers for Shellshock payloads.
	 *
	 * @param  Request $request  Normalised request object
	 * @param  array   $config   WAF configuration (reserved for future use)
	 * @return array{rule:string,matched:string,source:string,severity:string}|null
	 */
	public static function scan(Request $request, array $config): ?array
	{
		foreach ($_SERVER as $key => $value) {
			if (!str_starts_with($key, 'HTTP_')) {
				continue;
			}
			if (!is_string($value)) {
				continue;
			}
			$result = self::matchString($value, 'header');
			if ($result !== null) {
				return $result;
			}
		}

		return null;
	}

	// ------------------------------------------------------------------ //
	// Private helpers
	// ------------------------------------------------------------------ //

	private static function matchString(string $value, string $label): ?array
	{
		foreach (self::$rules as $rule) {
			if (preg_match($rule['pattern'], $value, $m)) {
				return [
					'rule'     => $rule['id'],
					'severity' => $rule['severity'],
					'matched'  => substr($m[0], 0, 120),
					'source'   => $label,
				];
			}
		}
		return null;
	}
}

File: aspnet/src/detectors/ShellshockDetector.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    /// <summary>
    /// Detects Shellshock (CVE-2014-6271 / CVE-2014-7169) exploitation attempts.
    /// Shellshock payloads are delivered via HTTP headers.
    /// </summary>
    public static class ShellshockDetector
    {
        private static readonly (string Id, string Severity, Regex Pattern)[] Rules =
        {
            ("shellshock-func",    "critical", new Regex(@"\(\s*\)\s*\{\s*[^}]*\}\s*;",  RegexOptions.Compiled)),
            ("shellshock-env-cmd", "critical", new Regex(@"\(\s*\)\s*\{\s*:;\s*\}\s*;",  RegexOptions.Compiled)),
        };

        public static DetectorResult Scan(WafRequest request)
        {
            foreach (var kv in request.Headers)
            {
                var r = MatchString(kv.Value, "header");
                if (r != null) return r;
            }
            return null;
        }

        private static DetectorResult MatchString(string value, string label)
        {
            foreach (var rule in Rules)
            {
                Match m = rule.Pattern.Match(value);
                if (m.Success)
                    return new DetectorResult(rule.Id, rule.Severity,
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value, label);
            }
            return null;
        }
    }
}
